cluster-autoscaler/cloudprovider/azure/azure_client.go

```diff
@@ -138,6 +145,143 @@ func (az *azDeploymentsClient) Delete(ctx context.Context, resourceGroupName, de
+		if err != nil {
+			klog.Errorf("NewClientSecretCredential failed: %v", err)
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("unsupported authorization method: %s", cfg.AuthMethod)
+	}
+	return cred, nil
+}
+
+func getAgentpoolClientRetryOptions(cfg *Config) azurecore_policy.RetryOptions {
+	if cfg.AuthMethod == authMethodCLI {
+		return azurecore_policy.RetryOptions{
+			MaxRetries: -1, // no retry when using CLI auth for UT
+		}
+	}
+	return azurecore_policy.RetryOptions{
+		MaxRetries:    5,
+		RetryDelay:    1 * time.Second,
+		MaxRetryDelay: 2 * time.Second,
+		TryTimeout:    10 * time.Second,
+	}
+}
+
+func newAgentpoolClientWithEndpoint(baseURL, subscriptionID string, retryOptions azurecore_policy.RetryOptions) (AgentPoolsClient, error) {
+	// for AKS managed CAS, we will use the ARMBaseURL to create a fake agent pool client
+	// so that the request will be sent to node provioner endpoint intead of the public ARM endpoint
+	agentPoolsClient, err := armcontainerservice.NewAgentPoolsClient(subscriptionID, nil,
+		&policy.ClientOptions{
+			ClientOptions: azurecore_policy.ClientOptions{
+				Cloud: cloud.Configuration{
+					Services: map[cloud.ServiceName]cloud.ServiceConfiguration{
+						cloud.ResourceManager: {
+							Endpoint: baseURL,
+							Audience: "UNKNOWN",
+						},
+					},
+				},
+				Telemetry: azurecore_policy.TelemetryOptions{
+					ApplicationID: userAgent,
+				},
+				Retry: retryOptions,
+			},
+		})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to init cluster agent pools client: %w", err)
+	}
+
+	klog.V(10).Infof("Successfully created agent pool client with ARMBaseURL")
+	return agentPoolsClient, nil
+}
+
+func newAgentpoolClient(cfg *Config) (AgentPoolsClient, error) {
+	retryOptions := getAgentpoolClientRetryOptions(cfg)
+
+	if cfg.ARMBaseURL != "" {
+		klog.V(10).Infof("Using ARMBaseURL to create agent pool client")
+		return newAgentpoolClientWithEndpoint(cfg.ARMBaseURL, cfg.SubscriptionID, retryOptions)
+	}
+
+	cred, err := getAgentpoolClientCredentials(cfg)
+	if err != nil {
+		klog.Errorf("failed to get agent pool client credentials: %v", err)
+		return nil, err
+	}
+
+	// default to public cloud
+	env := azure.PublicCloud
+	if cfg.Cloud != "" {
+		env, err = azure.EnvironmentFromName(cfg.Cloud)
+		if err != nil {
+			klog.Errorf("failed to get environment from name %s: with error: %v", cfg.Cloud, err)
+			return nil, err
+		}
+	}
+	agentPoolsClient, err := armcontainerservice.NewAgentPoolsClient(cfg.SubscriptionID, cred,
+		&policy.ClientOptions{
+			ClientOptions: azurecore_policy.ClientOptions{
+				Cloud: cloud.Configuration{
+					Services: map[cloud.ServiceName]cloud.ServiceConfiguration{
+						cloud.ResourceManager: {
+							Endpoint: env.ResourceManagerEndpoint,
+							Audience: env.TokenAudience,
+						},
+					},
+				},
+				Telemetry: azurecore_policy.TelemetryOptions{
+					ApplicationID: userAgent,
+				},
+				Retry: retryOptions,
+			},
+		})
```

Comment:
We are initializing the client twice here, can we share one fucntion for initalization and just pass in the right policy.ClientOptions? Also again would be good to use the base of the shared client options so that you dont' enouunter all the tcp bugs, throttling etc we encountered.